lib/src/lifecycle_module.dart

```diff
@@ -159,25 +206,80 @@ abstract class LifecycleModule {
   }
 
   /// Public method to trigger the Module unload cycle.
+  ///
   /// Calls shouldUnload(), and, if that completes successfully,
   /// continues to call onUnload() on the module and all registered child modules.
   /// If unloading is rejected, this method will complete with an error.
   Future unload() {
     Completer completer = new Completer();
-    ShouldUnloadResult canUnload = shouldUnload();
-    if (canUnload.shouldUnload) {
-      _willUnloadController.add(this);
-      Iterable<Future> unloadChildren = _childModules.map((c) => c.unload());
-      Future.wait(unloadChildren).then((_) {
-        _childModules.clear();
-        onUnload().then((_) {
-          _didUnloadController.add(this);
+    if (_isLoaded) {
+      ShouldUnloadResult canUnload = shouldUnload();
+      if (canUnload.shouldUnload) {
+        _willUnloadController.add(this);
+        Iterable<Future> unloadChildren = _childModules.map((c) => c.unload());
+        Future.wait(unloadChildren).then((_) {
+          _childModules.clear();
+          onUnload().then((_) {
+            _didUnloadController.add(this);
+            _isLoaded = false;
+            _isSuspended = false;
+            completer.complete();
+          });
+        });
+      } else {
+        // reject with shouldUnload messages
+        throw new ModuleUnloadCanceledException(canUnload.messagesAsString());
+      }
+    } else {
+      _logger.warning('Module "$name" is already unloaded, cannot unload.');
+      completer.complete();
+    }
+    return completer.future;
+  }
+
+  /// Public method to suspend the module.
+  ///
+  /// Suspend indicates to the module that it should go into a low-activity
+  /// state. For example, by disconnecting from backend services and unloading
+  /// heavy data structures.
+  Future suspend() {
```

Comment:
CP seems to put them in some kind of "logical" order so I tried to stick with that. Suspend comes before resume in that case because you always suspend before you can resume. When in Rome... I'm happy to change the order to whatever though, as long as it stays generally consistent throughout the library.